app/ota_client.py

```diff
@@ -526,48 +546,38 @@ def _create_symbolic_links(self, list_file, standby_path):
             slink.symlink_to(slinkf.srcpath)
             os.chown(slink, slinkf.uid, slinkf.gid, follow_symlinks=False)
 
-    def _set_statistics(self, processed_list):
-        st = self._statistics
-        st.files_processed = len(processed_list)
-
-        def set_st(op, files_processed, file_size_processed, elapsed_time):
-            _list = []
-            for i in processed_list[: st.files_processed]:
-                if i["op"] == op:
-                    i.pop("op")  # remove 'op' element
-                    _list.append(i)
-            setattr(st, files_processed, len(_list))
-
-            _list = _list if _list else [{}]  # add {} if empty
-            _total = reduce(operator.add, map(Counter, _list))
-            setattr(st, file_size_processed, _total.get("size", 0))
-            # convert from seconds to milli seconds as int
-            setattr(st, elapsed_time, int(_total.get("elapsed", 0) * 1000))
-            if op == "download":
-                st.errors_download = _total.get("errors", 0)
-
-        set_st(
-            "copy",
-            "files_processed_copy",
-            "file_size_processed_copy",
-            "elapsed_time_copy",
-        )
-        set_st(
-            "link",
-            "files_processed_link",
-            "file_size_processed_link",
-            "elapsed_time_link",
-        )
-        set_st(
-            "download",
-            "files_processed_download",
-            "file_size_processed_download",
-            "elapsed_time_download",
-        )
+    def _set_statistics(self, sts):
+        """
+        thread-safe modify statistics storage
+
+        st format:
+            {"size": int}  # file size
+            {"elapsed": int}  # elapsed time in seconds
+            {"op": str}  # operation. "copy", "link" or "download"
+            {"errors": int}  # number of errors that occurred when downloading.
+        """
+        all_processed = len(sts)
+        with self._statistics.acquire_staging_storage() as staging_storage:
+            already_processed = staging_storage.get("files_processed", 0)
+            if already_processed >= staging_storage.get("total_files", 0):
+                return
+
+            staging_storage["files_processed"] += all_processed - already_processed
+            for st in sts[already_processed:all_processed]:
+                _suffix = st.get("op")
+                if _suffix in {"copy", "link", "download"}:
+                    staging_storage[f"files_processed_{_suffix}"] += 1
+                    staging_storage[f"file_size_processed_{_suffix}"] += st.get(
+                        "size", 0
+                    )
+                    staging_storage[f"elapsed_time_{_suffix}"] += int(
+                        st.get("elapsed", 0) * 1000
+                    )
+                    staging_storage[f"errors_{_suffix}"] = st.get("errors", 0)
```

Comment:
I'd like to limit as follows and use increment(+=):
```suggestion
                    if _suffix == "download":
                        staging_storage[f"errors_{_suffix}"] += st.get("errors", 0)
```

Reply:
OK!